From 0e9fa15eb2e483fd1474bf737f25d1ae0aacdac0 Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Tue, 22 Jul 2014 14:39:12 -0400
Subject: [virt] kvm/nvmx: Do not inject NMI vmexits when L2 has a pending interrupt

Message-id: <1406040016-3289-158-git-send-email-pbonzini@redhat.com>
Patchwork-id: 86236
O-Subject: [RHEL7 PATCH v2 157/221] KVM: nVMX: Do not inject NMI vmexits when L2 has a pending interrupt
Bugzilla: 1116936
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Bandan Das <bsd@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jan Kiszka <jan.kiszka@siemens.com>

According to SDM 27.2.3, IDT vectoring information will not be valid on
vmexits caused by external NMIs. So we have to avoid creating such
scenarios by delaying EXIT_REASON_EXCEPTION_NMI injection as long as we
have a pending interrupt because that one would be migrated to L1's IDT
vectoring info on nested exit.

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
(cherry picked from commit 220c56729766444f3dd823f740a147ca6d82c4c6)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index e559675..2c9d21e 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -8176,7 +8176,8 @@ static int vmx_check_nested_events(struct kvm_vcpu *vcpu, bool external_intr)
  }
 
  if (vcpu->arch.nmi_pending && nested_exit_on_nmi(vcpu)) {
-  if (vmx->nested.nested_run_pending)
+  if (vmx->nested.nested_run_pending ||
+      vcpu->arch.interrupt.pending)
    return -EBUSY;
   nested_vmx_vmexit(vcpu, EXIT_REASON_EXCEPTION_NMI,
       NMI_VECTOR | INTR_TYPE_NMI_INTR |
-- 
1.7.1